1439 android client server preference

mark-prins opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #1439

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
Removed localStorage and have used the android SharedPreference instead.
Tried to use Capacitor Preferences but could not get the plugin to work 🙄 

It can be a little jumpy on the initial loads.. but does now work reliably. The client/server switch in Settings is working correctly too. I've also added the local server to the list of servers even when running in server mode - this is because the server has to exist in the list in order to connect to it.

![image](https://user-images.githubusercontent.com/9192912/228439343-d91a8af8-72ec-48e8-b0a3-0e0d3c8d47b9.png)

### Offroads
- Added the name of the server to the failed to connect error:
- The 'failed to connect' message is not shown if there was no previous server saved.
- Tweaked the layout of the screen to remove additional whitespace
- If the user is not logged in, then redirect to the login screen : otherwise the dashboard is shown, but as soon as you interact you are shown the logged out modal

# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->

Cleared data - ran in client mode, then checked that:

- [ ] you are taken to the server discovery screen & servers are available
- [ ] connecting to a remote server works
- [ ] after connecting to a remote server, turning off that server and restarting the android app shows the appropriate error
- [ ] reset again and connect to the local server, will take you to the initialisation screen
- [ ] Run various combinations of the above: closing the app, checking that you get back to what you expect, change client/server mode, clear data etc.

cleared data - ran in server mode

- [ ] taken to the initialisation screen ( this was taking you to the dashboard, though you aren't logged in )


## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
_No user facing changes_

andreievg commented on client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:
```diff
@@ -327,26 +330,25 @@ private NsdServiceInfo createLocalServiceInfo() {
     // manually adding this server should work
     private void addLocalServerToDiscovery() {
-        if(isAdvertising && isDiscovering && discoveredServers != null) {
+        if (isDiscovering && discoveredServers != null) {
```
So when we run as client we `do` want local server to be discovered ?

mark-prins replied:
I have reverted that - I thought that it was required, but cannot see why now after re-reading the code. will run a few more tests, but my view is that we don't need local when running in client mode.

andreievg commented on client/packages/common/src/hooks/useNativeClient/useNativeClient.ts:
```diff
@@ -139,7 +93,7 @@ export const useNativeClient = ({
     let connectToPreviousTimer: NodeJS.Timer | undefined = undefined;
 
-    if (autoconnect) {
+    if (autoconnect && !!state.previousServer) {
```
Great ! I also noticed this and had similar idea for fix, but that was when localStorage was `sync` method, not sure if this would always work now, since state.previousServer may not be set yet due to async nature of getting previousServer

mark-prins commented on client/package.json:
```diff
@@ -107,6 +107,7 @@
   },
   "dependencies": {
     "@bugsnag/js": "^7.16.2",
+    "@capacitor/core": "^4.7.1",
```
the core versions were inconsistent - I've moved to the root rather than common, since the plugin was using Capacitor, and have left it here. We've used Capacitor in `useNativeClient` and that's used in a number of packages so it makes sense to have it at the root.

mark-prins commented on client/packages/common/src/hooks/useNativeClient/useNativeClient.ts:
```diff
@@ -1,74 +1,32 @@
-  stopBarcodeScan: () => void;
-  readLog: () => Promise<{ log: string; error: string }>;
-}
+import {
```
the file was getting pretty long, so I broke it up. Hopefully it makes sense like this

mark-prins commented on client/packages/common/src/hooks/useNativeClient/helpers.ts:
```diff
@@ -0,0 +1,56 @@
+import { Capacitor, registerPlugin } from '@capacitor/core';
+import { FrontEndHost, NativeMode, NativeAPI } from './types';
+
+export const setPreference = async (
```
the preference methods are new - the rest of this is just moved from `useNativeClient`

mark-prins commented on client/packages/electron/src/electron.ts:
```diff
@@ -135,6 +136,16 @@ const start = (): void => {
   ipcMain.handle(IPC_MESSAGES.START_BARCODE_SCAN, () => barcodeScanner.start());
   ipcMain.on(IPC_MESSAGES.STOP_BARCODE_SCAN, () => barcodeScanner.stop());
+  // preference method stubs: not currently implemented for electron
+  ipcMain.on(IPC_MESSAGES.GET_PREFERENCE, (_event, preference: Preference) => ({
```
am thinking to implement with `electron-store` but thought to make a new issue for that

mark-prins commented on client/packages/host/src/Admin/Settings.tsx:
```diff
@@ -179,6 +181,10 @@ export const Settings: React.FC = () => {
       </>
     ) : null;
 
+  useEffect(() => {
```
this should be in `useNativeClient` but having the internal state in there wasn't working. The hook wasn't refreshing the value in the component that was using it. Since there's only two places that its used (here and `Android.tsx`), I have moved it out to those.

andreievg replied:
I think this would be the only real place that we would need to use `preferences` since url may not match discovery url, at this point I would just add url parameter to discovery page (that would trigger overwrite the localStorage)

mark-prins commented on client/packages/host/src/components/Android.tsx:
```diff
@@ -122,15 +135,16 @@ export const Android = () => {
     if (
       mode === NativeMode.Client &&
-      (!previousServer || connectToPreviousTimedOut)
+      (!previousServer?.ip || connectToPreviousTimedOut)
```
slightly clumsy - this is because there's no removePreference, and I'm setting the previousServer to an empty object when clearing it

andreievg commented on client/packages/common/src/hooks/useNativeClient/helpers.ts:
I like the file split, nicely structured (and useNativeClient was getting too big)

andreievg commented on client/packages/invoices/src/InboundShipment/DetailView/modals/InboundServiceLineEdit/useDraftServiceLines.ts:
```diff
@@ -46,12 +46,13 @@ export const useDraftServiceLines = () => {
-        const newPatch = { ...patch, totalAfterTax };
-        if (line.id === patch.id)
-          return { ...line, ...newPatch, isUpdated: true };
+        if (line.id === patch.id) {
```
I can't see description of this change, what should it do, how to test it ?

mark-prins replied:
It came from [this suggestion](https://github.com/openmsupply/open-msupply/pull/1311#discussion_r1141466686), which wasn't taken up by the dev before merging. I still think that it's a good idea, and functionally equivalent. 

To test - check that the total value is updated in the inbound shipment service line edit modal, when you make a change to the tax or amount values. The PR has more details.. perhaps too many details!

andreievg commented on package.json:
```diff
@@ -1,10 +1,10 @@
   "name": "open-msupply",
   "//": "Main version for the app, should be in semantic version format (any release candidate or test build should be separated by '-' i.e. 1.1.1-rc1 or 1.1.1-test",
-  "version": "1.1.6",
+  "version": "1.1.7",
```
do we do this in main or in develop ?

mark-prins replied:
true. usually only in develop if a migration is needed. will revert

andreievg commented on client/packages/host/src/Admin/Settings.tsx:
```diff
@@ -133,8 +134,9 @@ export const Settings: React.FC = () => {
 
-    localStorage.removeItem(PREVIOUS_SERVER_KEY);
-    setNativeMode(mode);
+    setPreference('previousServer', {} as FrontEndHost);
```
technically for async operations, something like below or in a new promise, etc.. not sure if it could be the case, but if setting preferences take longer then navigation would have an issue here

```
    (async () => {
      await setPreference('previousServer', {} as FrontEndHost);
      await setPreference('mode', mode);
      navigate(RouteBuilder.create(AppRoute.Android).build());
    })();
```

mark-prins replied:
thanks - that's much better

andreievg commented on client/packages/android/README.MD:
I think it's worth adding something to this file: `if you have "gradle" errors, try File ->  Sync Project with cradle files, in android studio`

mark-prins replied:
ah yes, good idea. it was in the block that I removed as part of the deps ugprade.

andreievg reviewed (changes requested):
Preference plugin works for me, but we do need to remember to add plugins to ExtendedWebViewClient, as per: https://github.com/openmsupply/open-msupply/compare/1439-android-client-server-preference...1439-example

As per README, this is somewhat tricky

```
Apart from usual local/packaged plugin setup for Capacitor, make sure to add plugin name to `onPageStarted` method in `ExtendedWebViewClient.java` (you can see what plugin names are available by using Android Studio debug and inspecting `bridge.plugins`). See point 2. in Capacitor Modifications (below).
```
![Add Prefrences plugin](https://user-images.githubusercontent.com/26194949/228691156-d342724a-c6e2-4010-a44b-47f9703d17f1.gif)


Also I think ideally we would need to overwrite localStorage (probably as a separate issue), for things like language etc, and to be confident that localStorage works uniformly across different apps, this would require LocalStorage to be rewritten, I wasn't sure why we would useListeners there rather then just zustand (which would be simpler). We would need to load all of the values from preferences/localStorage at the start, and await for them to be loaded (since local storage would be loaded in async manner)

The only required change, is trying to use the @capacitor/preferences plugin

mark-prins commented:
> As per README

I should have read the docs! 😆 
Thanks - that was a good pickup, and well explained. I've changed to use the Preferences, removed the manual implementation and retested it all. agree, that's much better. Changes in bb2f3962a 🙏

andreievg reviewed (approved):
Looks great, thanks for changes @mark-prins. 

I was wondering, maybe we need to add to carry over:
* For current Timor sites, when they upgrade to the future, they might get init screen ?
* Other localStorage things to use preferences ?

mark-prins commented:
> For current Timor sites, when they upgrade to the future, they might get init screen ?

yeah, I was unsure about adding a migration for them. They will get prompted to select 'server mode' again - though the data won't need to be synced. Nick thought it would be ok to have a one-off, click-the-button for them

> Other localStorage things to use preferences ?

I thought that for the other localStorage uses, it's probably ok to have them site-specific. It is unlikely that users will be changing the server, and the transient nature of the storage is probably ok for site customisations ( which columns to show, should the details panel / menu be opened ). The language might be annoying - but if set on the user account in central server then that's not an issue too.
